### PR TITLE
Userspace with some common keymaps.

### PR DESCRIPTION
**Pull request by GauthamYerroju**
~~Add my initial code userspace.~~

Userspace keymaps for some of my keyboards, with common features centralized in users/gotham.

## Description

Since I have a few keyboards now, I moved my common code to userspace. Borrowed a lot from drashna's userspace, and tested with the respective keyboards.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

None 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by GauthamYerroju** [comment]
Hey folks, any chance someone can take a look at this?

File: users/gotham/process_records.c [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            case KC_MAKE:
+                if (!record->event.pressed) {
+                    SEND_STRING("make " QMK_KEYBOARD ":" QMK_KEYMAP
+#if (defined(BOOTLOADER_DFU) || defined(BOOTLOADER_LUFA_DFU) || defined(BOOTLOADER_QMK_DFU))
+                                ":dfu BOOTLOADER=atmel-dfu"
+#elif defined(BOOTLOADER_HALFKAY)
+                                ":teensy BOOTLOADER=halfkay"
+#elif defined(BOOTLOADER_CATERINA)
+                                ":avrdude BOOTLOADER=caterina"
+#endif
```
Comment (drashna) [inline users/gotham/process_records.c:39]: you can replace this with `:flash`.  And then add the bootloader option based on the bootloader part. 

either way works, though.

Comment (GauthamYerroju) [reply users/gotham/process_records.c]: Ah, I missed that. Fixed.

File: keyboards/kyria/keymaps/gotham/config.h [hunk 2]
```diff
@@ -26,14 +25,26 @@
 
 #ifdef OLED_DRIVER_ENABLE
 #    define OLED_DISPLAY_128X64
-#    define OLED_TIMEOUT 10000
+
+#    ifdef OLED_CUSTOM_TIMEOUT
+#undef OLED_CUSTOM_TIMEOUT
+#define OLED_CUSTOM_TIMEOUT 30000
```
[the remark this answers is not in the document]
Comment (GauthamYerroju) [reply keyboards/kyria/keymaps/gotham/config.h]: Done!

**Comment by GauthamYerroju** [comment]
@drashna I've made a bunch of changes since opening this PR, and it occurs to me it might be bad form to push updates on an ongoing PR, sorry about that. I can close this and re-open another one if you'd like.

Thanks!

**Comment by drashna** [comment]
No, that's fine.  Especially for keymap PRs, we squash them, so it shows up as a single commit once it's been merged in.

File: keyboards/crkbd/keymaps/gotham/rules.mk [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+BACKLIGHT_ENABLE           = no  # Enable keyboard backlight functionality
+MIDI_ENABLE                = no  # MIDI controls
+AUDIO_ENABLE               = no  # Audio output on port C6
+UNICODE_ENABLE             = no  # Unicode
+BLUETOOTH_ENABLE           = no  # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE            = yes # Enable WS2812 RGB underlight
+SWAP_HANDS_ENABLE          = no  # Enable one-hand typing
+SLEEP_LED_ENABLE           = no  # Breathing sleep LED during USB suspend
+OLED_DRIVER_ENABLE         = yes # Enables the use of OLED displays
+LTO_ENABLE                 = yes # Enables link-time optimization (reduces firmware size)
```
Comment (drashna) [inline keyboards/crkbd/keymaps/gotham/rules.mk:18]:
```suggestion
BOOTMAGIC_ENABLE           = no  # Virtual DIP switch configuration
MOUSEKEY_ENABLE            = yes # Mouse keys
EXTRAKEY_ENABLE            = yes # Audio control and System control
CONSOLE_ENABLE             = no  # Console for debug
COMMAND_ENABLE             = no  # Commands for debug and configuration
NKRO_ENABLE                = no  # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
BACKLIGHT_ENABLE           = no  # Enable keyboard backlight functionality
MIDI_ENABLE                = no  # MIDI controls
AUDIO_ENABLE               = no  # Audio output on port C6
UNICODE_ENABLE             = no  # Unicode
BLUETOOTH_ENABLE           = no  # Enable Bluetooth with the Adafruit EZ-Key HID
RGBLIGHT_ENABLE            = yes # Enable WS2812 RGB underlight
SWAP_HANDS_ENABLE          = no  # Enable one-hand typing
SLEEP_LED_ENABLE           = no  # Breathing sleep LED during USB suspend
OLED_DRIVER_ENABLE         = yes # Enables the use of OLED displays
LTO_ENABLE                 = yes # Enables link-time optimization (reduces firmware size)
```

File: users/gotham/gotham.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#ifndef USERSPACE
+#define USERSPACE
```
Comment (drashna) [inline users/gotham/gotham.h:2]: This isn't needed.

```suggestion
```

File: users/gotham/gotham.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+void          matrix_init_keymap(void);
+void          keyboard_post_init_keymap(void);
+void          matrix_scan_keymap(void);
+void          suspend_power_down_keymap(void);
+void          suspend_wakeup_init_keymap(void);
+layer_state_t layer_state_set_keymap(layer_state_t state);
+layer_state_t default_layer_state_set_keymap(layer_state_t state);
+bool          led_update_keymap(led_t led_state);
+
+#endif
```
Comment (drashna) [inline users/gotham/gotham.h:53]:
```suggestion
```

File: users/gotham/transport_custom.c [hunk 6]
```diff
@@ -0,0 +1,343 @@
+#include <string.h>
+#include <stddef.h>
+
+#include "config.h"
```
Comment (drashna) [inline users/gotham/transport_custom.c:4]: Isn't needed.
```suggestion
```

**Comment by GauthamYerroju** [comment]
@drashna pushed your changes

**Comment by GauthamYerroju** [comment]
@drashna sorry about this. I missed your last comment about adding headers. Made a lot of changes since then (mostly convert most of my keyboards/keymaps/gotham/* to make use of userspace stuff). If you need me to break these into smaller PRs (one for userspace, and one for each keyboard, maybe), let me know.

Thanks a

**Comment by GauthamYerroju** [comment]
This has grown too much over the last few months, I'll break it into smaller PRs so I can properly test userspace changes on my keyboards.

**Pull request closed without merging**